drivers/i2s/i2s_cavs.c

```diff
@@ -0,0 +1,791 @@
+
+/* Device run time data */
+struct i2s_cavs_dev_data {
+	struct device *dev_dma;
+	struct stream tx;
+};
+
+#define DEV_NAME(dev) ((dev)->config->name)
+#define DEV_CFG(dev) \
+	((const struct i2s_cavs_config *const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct i2s_cavs_dev_data *const)(dev)->driver_data)
+
+static struct device *get_dev_from_dma_channel(u32_t dma_channel);
+static void dma_tx_callback(struct device *, u32_t, int);
+static void tx_stream_disable(struct stream *, struct ssp *const,
+				struct device *);
+
+static inline u16_t modulo_inc(u16_t val, u16_t max)
+{
+	val++;
+	return (val < max) ? val : 0;
+}
+
+/*
+ * Get data from the queue
+ */
+static int queue_get(struct ring_buf *rb, u8_t mode, void **mem_block,
+			size_t *size)
+{
+	unsigned int key;
+
+	key = irq_lock();
+
+	/* In case of ping-pong mode, the buffers are not freed after
+	 * reading. They are fixed in size. Another thread will populate
+	 * the pong buffer while the ping buffer is being read out and
+	 * vice versa. Hence, we just need to keep reading from buffer0
+	 * (ping buffer) followed by buffer1 (pong buffer) and the same
+	 * cycle continues.
+	 *
+	 * In case of non-ping-pong modes, each buffer is freed after it
+	 * is read. The tail pointer will keep progressing depending upon
+	 * the reads. The head pointer will move whenever there's a write.
+	 * If tail equals head, it would mean we have read everything there
+	 * is and the buffer is empty.
+	 */
+	if (rb->tail == rb->head) {
+		if ((mode & I2S_OPT_PINGPONG) == I2S_OPT_PINGPONG) {
+			/* Point back to the first element */
+			rb->tail = 0;
+		} else {
+			/* Ring buffer is empty */
+			irq_unlock(key);
+			return -ENOMEM;
+		}
+	}
+
+	*mem_block = rb->buf[rb->tail].mem_block;
+	*size = rb->buf[rb->tail].size;
+	rb->tail = modulo_inc(rb->tail, rb->len);
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+/*
+ * Put data in the queue
+ */
+static int queue_put(struct ring_buf *rb, u8_t mode, void *mem_block,
+			size_t size)
+{
+	u16_t head_next;
+	unsigned int key;
+
+	key = irq_lock();
+
+	head_next = rb->head;
+	head_next = modulo_inc(head_next, rb->len);
+
+	/* In case of ping-pong mode, the below comparison incorrectly
+	 * leads to complications as the buffers are always predefined.
+	 * Hence excluding ping-pong mode from this comparison.
+	 */
+	if ((mode & I2S_OPT_PINGPONG) != I2S_OPT_PINGPONG) {
+		if (head_next == rb->tail) {
+			/* Ring buffer is full */
+			irq_unlock(key);
+			return -ENOMEM;
+		}
+	}
+
+	rb->buf[rb->head].mem_block = mem_block;
+	rb->buf[rb->head].size = size;
+	rb->head = head_next;
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int start_dma(struct device *dev_dma, u32_t channel,
+		     struct dma_config *cfg, void *src, void *dst,
+		     u32_t blk_size)
+{
+	struct dma_block_config blk_cfg;
+	int ret;
+
+	memset(&blk_cfg, 0, sizeof(blk_cfg));
+	blk_cfg.block_size = blk_size;
+	blk_cfg.source_address = (u32_t)src;
+	blk_cfg.dest_address = (u32_t)dst;
```

Comment:
Style; doing this with a struct literal allows the compiler to generate better code and is more readable:

    blk_cfg = {
       .block_size = blk_size,
       .source_address = (u32_t)src,
       .dest_address = (u32_t)dst,
    }

Reply:
sure. will make the change.